src/common/utils/fold.h

```diff
@@ -0,0 +1,57 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#ifndef SRC_COMMON_UTILS_FOLD_H_
+#define SRC_COMMON_UTILS_FOLD_H_
+
+// This file contains a C++ implementation of the classic "fold" and
+// "right_fold" higher-order functions. This is somewhat redundant with
+// std::accumulate; unfortunately, std::accumulate requires its accumulator to
+// be copyable until C++20.
+//
+// Note that "right_fold" is named "rfold". This is a bit of a pun: if you come
+// from a functional languages background, the "r" naturally expands to "right",
+// while if you are fluent with C++ iterators, "r" naturally means "reverse".
+// In this context it actually means both: a right_fold achieved by using
+// reverse iterators on the provided container.
+
+#include <iterator>
+
+namespace raksha::common::utils {
+
+template <class I, class A, class F>
+A fold_iter(I begin_iter, I end_iter, A accumulated, F fold_fn) {
```

Comment:
I was thinking that it might be useful to expose it, as this allows you to `fold` a wider variety of things. For instance, with `fold_iter`, you can `fold` over regular old pointers on a contiguous chunk of memory. It also corresponds very directly with `std::accumulate`, so a transition to C++20 could involve just making the body of this function a call to `std::accumulate`.

Reply:
Added a test and comment for this function, as one should do for fns in the public API.

Reply:
just call it `fold` then?

Reply:
Done.